### ISSUE TITLE
Bilby Prior confusion and possible solution

### ISSUE THREAD

**Issue by EthanMarx** [ML4GW/BBHNet#271]
I'm realizing the source frame / detector frame confusion stems from the fact that we use `bilby` priors for two distinct purposes:

1. To sample from for generating injections in the detector frame to train our network. This is what the `bilby` priors were built for since performing mcmc measures detector frame quantities.
2. For measuring a sensitive distance from a given population (i.e. generating our testing set). We care about the astrophysical population, and thus these priors should be source frame based. However, injections are created with detector frame quantities, and thus we require a conversion before generating the waveforms. This causes problems in our current `timeslides` implementation [here](https://github.com/ML4GW/BBHNet/blob/main/projects/sandbox/datagen/datagen/utils/timeslides.py#L18). 

I think this simple wrapper that converts masses to detector frame is the easiest way to get around these problems with the least amount of code changes. Open to thoughts and suggestions. 
```python
class PopulationPrior():
    def __init__(self, population: bilby.core.prior.PriorDict):
        self.population = population
        
    def sample_detector_frame(self, n_samples: int):
        samples = self.population.sample(n_samples)
        # convert masses to detector frame
        samples["mass_1"] = samples["mass_1"] * (1 + samples["redshift"])
        samples["mass_2"] = samples["mass_2"] * (1 + samples["redshift"])
        return samples
```

**Comment by alecgunny** [comment]
Wouldn't we want to use the same distribution for both training and testing? Ultimately the test distribution is the one we want to optimize for, so we may as well do it directly

**Comment by alecgunny** [comment]
Also been playing with the `PriorDict` class and it looks like just inheriting and overriding the `sample` method makes all the `dict` behavior work as intended (in that they return the subclassed object, at least for in-place methods like `.update` and `__getitem__`, haven't tested against the new functionality like using `|`), so could do something like
```python
from typing import Optional
from bilby.core.prior improt PriorDict as BilbyPriorDict


class PriorDict(BilbyPriorDict):
    def sample(self, n: Optional[int] = None, source_frame: bool = False):
        samples = super().sample(n)
        if not source_frame:
            return samples
        # do the key replacement here
        return samples
```
Of course if it turns out we just want to do everything in source frame, we can ditch the kwarg and just do the replacement always (maybe checking to make sure the mass keys exist)

**Comment by EthanMarx** [comment]
Thats a very reasonable suggestion. I go back and forth about this point. The distribution we inject for testing is ultimately re-weighted to an entirely different distribution when we calculate our metrics. In the O3 catalog paper the pipelines don't actually calculate a VT for the entire rates and pop distribution, they perform the population re-weighting to the specific mass bins. If we truly want to optimize for the "test" distribution should we then only train (and completely overfit) to these narrow mass bins? 

I guess my philosophy is that because we don't know what the universe will throw at us in the future, we should remain as agnostic as possible. As I am typing this, I recall that some of the search pipelines may incorporate population priors into their ranking statistics. We should definitely look into this.

**Comment by EthanMarx** [comment]
I like the subclass idea as well. For both implementations we will have to be careful about utilizing the other `PriorDict` methods like `.prob`, which will require a jacobian.

**Comment by alecgunny** [comment]
Definitely agree you want the distribution for which your network is optimized to be as robust as possible in order to account for uncertainties in the parameters of your distribution, and the differences we're talking about here are probably too minor to result in massively different behavior from the network anyway.

I guess my point is that we're ultimately going to score ourselves on the predictions the network produces on our test set, regardless of how you re-weight the contribution of those predictions after the fact. So since we _can_ optimize for making the best predictions possible on the distribution of that test set (by using it during training), I don't really see any reason _not_ to. Like ultimately we have to pick a training distribution, so we may as well pick one that matches the one we'll use to produce predictions later, especially since I don't know what the physical motivation would be to use the unshifted prior.
